### PR TITLE
Improve the cache layer in application management module

### PR DESCRIPTION
**Pull request by senthalan**
### Proposed changes in this pull request

Fix : https://github.com/wso2/product-is/issues/4639

### When should this PR be merged

ASAP

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Comment by senthalan** [comment]
Closing this, as we want improvement on the caching layer.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        this.appDAO = appDAO;
+        appCacheByName = IdentityServiceProviderCache.getInstance();
+        appCacheByInboundAuth = ServiceProvideCacheInboundAuth.getInstance();
+        appCacheByID = ServiceProviderCacheID.getInstance();
+    }
+
+    public ServiceProvider getApplication(String applicationName, String tenantDomain) throws
+            IdentityApplicationManagementException {
+
+        ServiceProvider serviceProvider = getApplicationFromCache(applicationName, tenantDomain);
```
Comment (darshanasbg) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java]: add debug log to indicate a cache hit or a miss

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            ApplicationMgtUtil.startTenantFlow(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+
+            IdentityServiceProviderCacheKey nameKey = new IdentityServiceProviderCacheKey(serviceProvider
+                    .getApplicationName(), tenantDomain);
+            IdentityServiceProviderCacheEntry nameEntry = new IdentityServiceProviderCacheEntry(serviceProvider);
+            appCacheByName.addToCache(nameKey, nameEntry);
+
+            ServiceProviderIDCacheKey idKey = new ServiceProviderIDCacheKey(serviceProvider.getApplicationID());
+            ServiceProviderIDCacheEntry idEntry = new ServiceProviderIDCacheEntry(serviceProvider);
+            appCacheByID.addToCache(idKey, idEntry);
```
Comment (darshanasbg) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java]: cant we add to ServiceProvideCacheInboundAuth?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/cache/ServiceProvideCacheInboundAuth.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.wso2.carbon.identity.application.mgt.internal.cache;
+
+import org.wso2.carbon.identity.application.common.cache.BaseCache;
+
+/**
+ * Cache to maintain the clint id - service provider name.
+ */
+public class ServiceProvideCacheInboundAuth extends
```
Comment (darshanasbg) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/cache/ServiceProvideCacheInboundAuth.java]: Typo in the class name, missing letter 'r' in Provider word.

Comment (darshanasbg) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/cache/ServiceProvideCacheInboundAuth.java]: Better to have Cache part at the end of the class name.
ie. ServiceProvideCacheInboundAuth ->
ServiceProviderByInboundAuthCache
ServiceProviderByIdCache
etc.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java (hunk 1 above)
Comment (senthalan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java]: added with https://github.com/wso2/carbon-identity-framework/pull/2112/commits/028b538c9deb0987b027a77d46f237652f4eeabc

**Pull request merged**